mesonbuild/dependencies/dev.py

```diff
@@ -700,6 +702,110 @@ def __init__(self, environment: 'Environment', kwargs: JNISystemDependencyKW):
+        # The only place where the full version is stored, seems to be the Version field in msdia*.dll resources.
+        dllname = os.path.basename(self.dll)
+        versionstr = dllname[len('msdia'):-len('.dll')]
+        if versionstr[-1] == '0':
+            self.version = versionstr[:-1]
+        else:
+            mlog.error(f"Unexpected DIA SDK version string in '{dllname}'")
+            self.version = 'unknown'
+
+        self.compile_args.append('-I' + include)
+        self.link_args.append(lib)
+        self.is_found = True
+        return True
+
+    # Check if compiler has a built-in macro defined
+    @staticmethod
+    def _has_define(compiler: 'Compiler', dname: str, env: 'Environment') -> bool:
+        defval, _ = compiler.get_define(dname, '', env, [], [])
+        return defval is not None
+
+    def __init__(self, environment: 'Environment', kwargs: T.Dict[str, T.Any]) -> None:
+        super().__init__('diasdk', environment, kwargs)
+        self.is_found = False
+
+        compilers = environment.coredata.compilers.host
+        if 'cpp' in compilers:
+            compiler = compilers['cpp']
+        elif 'c' in compilers:
+            compiler = compilers['c']
+        else:
+            mlog.error('DIA SDK is only supported in C and C++ projects')
+            return
+
+        is_msvc_clang = compiler.id == 'clang' and self._has_define(compiler, '_MSC_VER', environment)
+        if compiler.id not in {'msvc', 'clang-cl'} and not is_msvc_clang:
+            mlog.error('DIA SDK is only supported with Microsoft Visual Studio compilers')
+            return
+
+        cpu_translate = {'arm': 'arm', 'aarch64': 'arm64', 'x86': '.', 'x86_64': 'amd64'}
+        cpu_family = environment.machines.host.cpu_family
+        cpu = cpu_translate.get(cpu_family)
+        if cpu is None:
+            mlog.error(f'DIA SDK is not supported for "{cpu_family}" architecture')
+            return
+
+        vsdir = os.environ.get('VSInstallDir')
+        if vsdir is None:
+            mlog.error("Environment variable VSInstallDir required for DIA SDK is not set")
+            return
+
+        diadir = os.path.join(vsdir, 'DIA SDK')
+        if self._try_path(diadir, cpu):
+            mlog.debug('DIA SDK was found at default path: ', diadir)
+            self.is_found = True
+            return
+        mlog.debug('DIA SDK was not found at default path: ', diadir)
+
+        return
+
+    def get_variable(self, *, cmake: T.Optional[str] = None, pkgconfig: T.Optional[str] = None,
+                     configtool: T.Optional[str] = None, internal: T.Optional[str] = None,
+                     default_value: T.Optional[str] = None,
+                     pkgconfig_define: PkgConfigDefineType = None) -> str:
+        if internal == 'dll':
+            return self.dll
+        if default_value is not None:
+            return default_value
+        raise DependencyException(f'Cannot get variable {internal}')
```

Comment:
I... get what this is trying to do, but it still seems weird. :D

@dcbaker this is a *system* dependency piggybacking onto the mechanism for `declare_dependency(variables: [...])` support?

Do we want to extend the get_variable interface to cope with system dependencies? consider them equivalent to configtool or internal? Simply expect them to use get_variable('dll') without a type-specific request?

Reply:
I would prefer to extend with `system` type. I still think the no type selection thing is an anti-feature, it lulls people into a false sense of "everything will work right all the time!" and it only happens to work in the case of pkg-config and wraps.